supervision/metrics/core.py

```diff
@@ -0,0 +1,210 @@
+            assert len(content_1) == len(class_ids_1)
+            for class_id in set(class_ids_1):
+                content_of_class = content_1[class_ids_1 == class_id]
+                if class_id not in self._data_1:
+                    self._data_1[class_id] = content_of_class
+                    continue
+                self._data_1[class_id] = np.vstack(
+                    (self._data_1[class_id], content_of_class)
+                )
+
+        if content_2 is not None and len(content_2) > 0:
+            assert len(content_2) == len(class_ids_2)
+            for class_id in set(class_ids_2):
+                content_of_class = content_2[class_ids_2 == class_id]
+                if class_id not in self._data_2:
+                    self._data_2[class_id] = content_of_class
+                    continue
+                self._data_2[class_id] = np.vstack(
+                    (self._data_2[class_id], content_of_class)
+                )
+
+    def __iter__(
+        self,
+    ) -> Iterator[Tuple[int, Optional[npt.NDArray], Optional[npt.NDArray]]]:
+        class_ids = sorted(
+            set.union(set(self._data_1.keys()), set(self._data_2.keys()))
+        )
+        for class_id in class_ids:
+            yield (
+                class_id,
+                self._data_1.get(class_id, None),
+                self._data_2.get(class_id, None),
+            )
+
+    def _get_content(
+        self, data: Union[npt.NDArray, Detections]
+    ) -> Optional[npt.NDArray]:
+        """Return boxes, masks or oriented bounding boxes from the data."""
+        if not isinstance(data, (Detections, np.ndarray)):
+            raise ValueError(
+                f"Invalid data type: {type(data)}. Only Detections or np.ndarray are supported."
+            )
+        if isinstance(data, np.ndarray):
+            return data
+
+        if self._metric_target == MetricTarget.BOXES:
+            return data.xyxy
+        if self._metric_target == MetricTarget.MASKS:
+            return data.mask
+        if self._metric_target == MetricTarget.ORIENTED_BOUNDING_BOXES:
+            obb = data.data.get(config.ORIENTED_BOX_COORDINATES, None)
+            if isinstance(obb, list):
+                obb = np.array(obb, dtype=np.float32)
+            return obb
+        raise ValueError(f"Invalid metric target: {self._metric_target}")
+
+    def _get_class_ids(
+        self, data: Union[npt.NDArray, Detections]
+    ) -> npt.NDArray[np.int_]:
+        if self._class_agnostic or isinstance(data, np.ndarray):
+            return np.array([CLASS_ID_NONE] * len(data), dtype=int)
+        assert isinstance(data, Detections)
+        if data.class_id is None:
+            return np.array([CLASS_ID_NONE] * len(data), dtype=int)
+        return data.class_id
+
+    def _validate_class_ids(self, class_id: npt.NDArray[np.int_]) -> None:
+        class_set = set(class_id)
+        if len(class_set) >= 2 and -1 in class_set:
```

Comment:
1.
Changed to `CLASS_ID_NONE`, exposed it in `__init__` and moved the docstring where it would apply to the `CLASS_ID_NONE` variable.

Here's the broader question. Do we try to always return a dict `dict[int, MetricSpecificResult]`, or return only `MetricSpecificResult` when using class-agnostic metrics? Packages like torchmetrics chose the latter, returning the result only, if `class_agnostic` is set.

I'm a proponent of less ambiguity in return types. The question is, whether the following is cumbersome or not:

```python
detections_1 = ...
detections_2 = ...

iou_metric = sv.metrics.IntersectionOverUnion(class_agnostic=True)
iou_metric.update(detections_1, detections_2)

ious = iou_metric.compute()[sv.metrics.CLASS_ID_NONE]
# or
ious = iou_metric.compute()[None]
# or
ious = iou_metric.compute()[-1]

# versus

ious = iou_metric.compute()
```

Case 1 is cumbersome unless we can reduce size of `CLASS_ID_NONE` or put it in global scope.
Case 2 I find rather clear, but people will need to check their keys for None or ignore a lot of mypy squiggles. It may also clash with 0.
Case 3 feels unnatural, as it seems you're popping off the end of the array.

Case 4 is cleanest, but will always need an extra check to determine which of the 2 Union outputs was returned.

That's the reason for looking at case 1 and having `CLASS_ID_NONE`. It gives a fairly reasonable API, even if verbose.

Reply:
2.
> I'm also not sure what are we checking here?

Good catch, the input should have been formed of both metrics. I did it inside the function now.
I'm still thinking of this one.